enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/scenarios/ExceptionSuppressor.java

```diff
@@ -0,0 +1,77 @@
+package org.neo4j.causalclustering.scenarios;
+
+import org.neo4j.function.ThrowingAction;
+
+/**
+ * Allows clean error handling when closing down multiple resources which are independent.
+ * One can differentiate exceptions to be caught and suppressed from those which should cause
+ * the current chain of suppressed exceptions to be thrown.
+ * <p>
+ * On {@link #close()} the chain of suppressed exceptions is thrown.
+ */
+class ExceptionSuppressor implements AutoCloseable
+{
+    private final Class defaultSuppressedClass;
+    private Throwable first = null;
+
+    public ExceptionSuppressor()
+    {
+        this( Exception.class );
+    }
+
+    public ExceptionSuppressor( Class defaultSuppressedClass )
+    {
+        this.defaultSuppressedClass = defaultSuppressedClass;
+    }
+
+    public <E extends Throwable> void execute( ThrowingAction<E> action, Class suppressedClass ) throws E
+    {
+        try
+        {
+            action.apply();
+        }
+        catch ( Throwable ex )
+        {
+            if ( first == null )
+            {
+                first = ex;
+            }
+            else
+            {
+                first.addSuppressed( ex );
```

Comment:
I'd suggest adding `first != ex` check because this is a generic utility class that does not know which actions it execute. Same exception might be thrown by different actions and self-suppressions are not allowed.